docs/config.json.md
# Config.json explanation

## In most cases, you only need to modify

- AutoMining、
- PayToAddr、
- Magic、
- SeedList、
- HttpInfoPort、
- HttpRestPort、
- HttpWsPort、
- HttpJsonPort、
- NodePort，
- ActiveNet  //for reduce the mining blocks interval
## Inline Explanation

```JSON
{
  "Configuration": {
    "Magic": 20180312,      //Magic Number：Segregation for different subnet. No matter the port number, as long as the magic number not matching, nodes cannot talk to each others.
    "Version": 23,          //Version number
    "SeedList": [           //SeedList. Other nodes will look up this seed list to connect to any of those seed in order to get all nodes addresses.
      "127.0.0.1:10338",    //At least one seed in this list. Format is "IP address : Port"
      "127.0.0.1:20338",
      "127.0.0.1:30338",
      "127.0.0.1:40338"
    ],
    "HttpInfoPort": 10333,  //Local web portal port number. User can go to http://127.0.0.1:10333/info to access the web UI
    "HttpInfoStart": true,  //true to start the webUI, false to disable
    "HttpRestPort": 10334,  //Restful port number
    "HttpWsPort": 10335,    //Websocket port number
    "WsHeartbeatInterval": 60,
    "HttpJsonPort": 10336,  //RPC port number
    "NodePort": 10338,      //P2P port number
    "NodeOpenPort": 10866,  //P2P port number for open service
    "OpenService": true,    //true to enable open service, false to disable
    "PrintLevel": 1,        //Log level. Level 0 is the highest, 6 is the lowest.
    "MaxLogsSize": 5000,    //Max total logs size in MB
    "MaxPerLogSize": 20,    //Max per log file size in MB
    "IsTLS": false,         //TLS connection, true or false
    "CertPath": "./sample-cert.pem",  //Certificate path
    "KeyPath": "./sample-cert-key.pem",
    "CAPath": "./sample-ca.pem",
    "MultiCoreNum": 4,      //Max number of CPU cores to mine ELA
    "MaxTransactionInBlock": 10000, //Max transaction number in each block
    "MaxBlockSize": 8000000,        //Max size of a block
    "MinCrossChainTxFee": 10000,    //Minimal cross-chain transaction fee
    "PowConfiguration": {           //
      "PayToAddr": "",              //Pay bonus to this address. Cannot be empty if AutoMining set to "true".
      "AutoMining": false,          //Start mining automatically? true or false
      "MinerInfo": "IOEX",           //No need to change.
      "MinTxFee": 100,              //Minimal mining fee
      "ActiveNet": "MainNet"        //Network type. Choices: MainNet、TestNet、RegNet，RegNet. Mining interval are 120s、10s、1s accordingly. Difficulty factor high to low.
    }
  }
}

```


docs/IOEX_Wallet_Development_Environment_CN.md
# IOEX Wallet Development Environment Document

## 開發伺服器獲取

1. 使用郵箱將 `ssh public key` 發送給 `service@ioex.co` 請求分配開發伺服器，管理員分配成功後回復相應的伺服器位址和存取方法。

    通常推薦使用證書方式登錄伺服器，如果需要使用密碼登錄也可以在郵件中進行說明。

2. 根據需求後會給您分配兩台伺服器：`鏈節點伺服器` 和 `錢包伺服器`。

## IOEX節點服務

### 環境說明

1. 應用程式預設放置在 `/opt/ioex_node` 目錄中，一個運行中的包含命令列用戶端的節點目錄內容如下：

    ```bash
    ls -al /opt/ioex_node

    total 24924
    drwxrwxr-x 4 ioex ioex     4096 Apr 20 17:19 ./
    drwxrwxr-x 5 ioex ioex     4096 Apr 20 11:50 ../
    drwxr-xr-x 2 ioex ioex     4096 Apr 20 12:08 Chain/
    -rw-rw-r-- 1 ioex ioex       34 Apr 20 11:51 cli-config.json
    -rw-rw-r-- 1 ioex ioex     1088 Apr 20 12:02 config.json
    -rwxrwxr-x 1 ioex ioex 13565344 Apr 20 11:51 ioex-cli*
    -rw-rw-r-- 1 ioex ioex      444 Apr 20 12:01 keystore.dat
    drwxrw-r-- 2 ioex ioex     4096 Apr 20 12:07 Log/
    -rwxrwxr-x 1 ioex ioex 11648196 Apr 20 11:50 ioex*
    -rw-r--r-- 1 ioex ioex   274432 Apr 20 17:19 wallet.db
    ```

   * `ioex` 是鏈節點的啟動程式，相應的設定檔為 `config.json`。

   * `ioex-cli` 是一個命令列管理工具，可以用來創建錢包，轉帳和查看區塊資料，設定檔為`cli-config.json`。

   * `keystore.dat` 和 `wallet.db` 是錢包的秘鑰和錢包數據。

   * `Chain` 和 `Log` 目錄是區塊鏈帳本資料以及日誌資料。

   * 鏈節點和命令列用戶端代碼和編譯方法以及配置和命令列參數可以參考：

     * [ioeX.MainChain](https://github.com/ioeXNetwork/ioeX.MainChain/blob/master/README.md)
     * [ioeX.Client](https://github.com/ioeXNetwork/ioeX.Client/blob/master/README.md)

2. 預設情況下分配給你的節點伺服器會連接到一個專門用於開發的一組節點伺服器，自動挖礦並且把所獲取的幣發送到本地的錢包中；你可以參考 [ioeX.Client](https://github.com/ioeXNetwork/ioeX.Client/blob/master/README.md) 中的相關查詢命令進行查詢或者其他操作。

   備註：部分命令列操作需要提供一個錢包密碼，預設是 `ioex`。

3. 節點伺服器使用如下埠提供服務：

    ```bash
    2[*]334  提供 Http Rest api
    2[*]335  提供 Web Socket api
    2[*]336  用於命令列用戶端連接到節點獲取資料
    2[*]338  提供節點之間的資料同步
    ```

   備註：`*` 表示這個數字不確定，會在成功分配伺服器後確定。

### Node API

可參考 API示例：

* [IOEX_Wallet_Node_API](IOEX_Wallet_Node_API_CN.md)

## IOEX錢包服務

1. IOEX錢包服務包含兩個前臺的服務：`browser` 和 `wallet`；以及另外三個後台工作的服務 `worker`、 `api`、 `notification`。

2. 服務簡介：

    ```bash
    browser:  區塊流覽器使用者介面，可以查看區塊資料，查詢交易
    wallet:   web錢包使用者介面，可以進行錢包創建，資產查詢，轉帳操作
    worker:   從節點獲取區塊資訊，寫入緩存資料庫
    api:      為 wallet 提供介面
    notification:
    ```

3. 上面的這些服務的應用程式都在 `/data/www/` 目錄下並使用 `xxxx.test.ioex.org` 作為目錄名稱；目前所有服務都是使用 `supervisor` 進行管理，相關設定檔在 `/etc/supervisor/conf.d` 目錄下，可以使用 `supervisorctl [start|stop|restart] <service_name>` 來啟動或者停止某個服務：

    ```bash
    supervisorctl stop worker.test.ioex.org

    supervisorctl restart all
    ```

   使用 `supervisor` 管理服務的詳細文檔可以在 [這裡](http://www.supervisord.org/) 查看。

4. IOEX錢包的 `web` 伺服器使用 `nginx`, 相關設定檔在 `/etc/nginx/conf.d` 目錄。

   開發環境的 `https` 服務使用了自簽名的證書，所以使用流覽器進行訪問時需要手動添加一下對證書的信任，另外分配的開發環境默認不會使用標準的`80` `443`埠，所以訪問時候要加上相應的埠。


config/config.go
package config

import (
	"bytes"
	"encoding/json"
	"io/ioutil"
	"log"
	"math/big"
	"os"
	"time"
)

const (
	DefaultConfigFilename = "./config.json"
	MINGENBLOCKTIME       = 2
	DefaultGenBlockTime   = 6
)

var (
	Parameters configParams
	Version    string
	mainNet    = &ChainParams{
		Name:                "MainNet",
		PowLimit:            new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), 255), big.NewInt(1)),
		PowLimitBits:        0x1e00f8ff,
		TargetTimePerBlock:  time.Minute * 2,
		TargetTimespan:      time.Minute * 2 * 720,
		AdjustmentFactor:    int64(4),
		MaxOrphanBlocks:     10000,
		MinMemoryNodes:      20160,
		CoinbaseLockTime:    100,
		ChainStorePath:      "Chain",
		FoundationAddress1:  "EeKhwVwcHudP12cMmHY7FyZKmhawd4NyMv",
		FoundationAddress2:  "EWKRQLkd1WtAK2Y6WUHKqefHwDjAyGxTEW",
		FoundationAddress3:  "Ebkho6hHKZdwhoFqC5rCfA3YvSV5TC5D7U",
		FoundationAddress4:  "EWbWYW5qfsH8ZEXDiPfLJ2H5N1AC9BKU7R",
		FoundationAddress5:  "EaJXQytPHwUZF11NWezBgHr9CzsWW5gRy7",
		FoundationAddress6:  "EeC9gKGoAuYwN1eGyC4XV3BR6HqVwQa9Ci",
		FoundationAddress7:  "EMTUBvw8ALSNtYU8VwfdqPLCsDCT1GXbii",
		FoundationAddress8:  "EPivxnWofDfeCSAmmS9pKdYBxi9FyUVnR5",
		FoundationAddress9:  "EWgdFRoPeZnv1VrGjGev1XGHKWxU8wsqAM",
		FoundationAddress10: "EN2QYLxznYT8MaciT2MkNfs2TcMog72h7a",
		FoundationAddress11: "ELPp9Gr1qvTjXEGLWVbYG288A9fvxeyXZJ",
	}
	testNet = &ChainParams{
		Name:                "TestNet",
		PowLimit:            new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), 255), big.NewInt(1)),
		PowLimitBits:        0x1e1da5ff,
		TargetTimePerBlock:  time.Second * 10,
		TargetTimespan:      time.Second * 10 * 10,
		AdjustmentFactor:    int64(4),
		MaxOrphanBlocks:     10000,
		MinMemoryNodes:      20160,
		CoinbaseLockTime:    100,
		ChainStorePath:      "Chain/TestNet",
		FoundationAddress1:  "EeKhwVwcHudP12cMmHY7FyZKmhawd4NyMv",
		FoundationAddress2:  "EWKRQLkd1WtAK2Y6WUHKqefHwDjAyGxTEW",
		FoundationAddress3:  "Ebkho6hHKZdwhoFqC5rCfA3YvSV5TC5D7U",
		FoundationAddress4:  "EWbWYW5qfsH8ZEXDiPfLJ2H5N1AC9BKU7R",
		FoundationAddress5:  "EaJXQytPHwUZF11NWezBgHr9CzsWW5gRy7",
		FoundationAddress6:  "EeC9gKGoAuYwN1eGyC4XV3BR6HqVwQa9Ci",
		FoundationAddress7:  "EMTUBvw8ALSNtYU8VwfdqPLCsDCT1GXbii",
		FoundationAddress8:  "EPivxnWofDfeCSAmmS9pKdYBxi9FyUVnR5",
		FoundationAddress9:  "EWgdFRoPeZnv1VrGjGev1XGHKWxU8wsqAM",
		FoundationAddress10: "EN2QYLxznYT8MaciT2MkNfs2TcMog72h7a",
		FoundationAddress11: "ELPp9Gr1qvTjXEGLWVbYG288A9fvxeyXZJ",
	}
	regNet = &ChainParams{
		Name:                "RegNet",
		PowLimit:            new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), 255), big.NewInt(1)),
		PowLimitBits:        0x207fffff,
		TargetTimePerBlock:  time.Second * 1,
		TargetTimespan:      time.Second * 1 * 10,
		AdjustmentFactor:    int64(4),
		MaxOrphanBlocks:     10000,
		MinMemoryNodes:      20160,
		CoinbaseLockTime:    100,
		ChainStorePath:      "Chain/RegNet",
		FoundationAddress1:  "EeKhwVwcHudP12cMmHY7FyZKmhawd4NyMv",
		FoundationAddress2:  "EWKRQLkd1WtAK2Y6WUHKqefHwDjAyGxTEW",
		FoundationAddress3:  "Ebkho6hHKZdwhoFqC5rCfA3YvSV5TC5D7U",
		FoundationAddress4:  "EWbWYW5qfsH8ZEXDiPfLJ2H5N1AC9BKU7R",
		FoundationAddress5:  "EaJXQytPHwUZF11NWezBgHr9CzsWW5gRy7",
		FoundationAddress6:  "EeC9gKGoAuYwN1eGyC4XV3BR6HqVwQa9Ci",
		FoundationAddress7:  "EMTUBvw8ALSNtYU8VwfdqPLCsDCT1GXbii",
		FoundationAddress8:  "EPivxnWofDfeCSAmmS9pKdYBxi9FyUVnR5",
		FoundationAddress9:  "EWgdFRoPeZnv1VrGjGev1XGHKWxU8wsqAM",
		FoundationAddress10: "EN2QYLxznYT8MaciT2MkNfs2TcMog72h7a",
		FoundationAddress11: "ELPp9Gr1qvTjXEGLWVbYG288A9fvxeyXZJ",
	}
)

type PowConfiguration struct {
	PayToAddr  string `json:"PayToAddr"`
	AutoMining bool   `json:"AutoMining"`
	MinerInfo  string `json:"MinerInfo"`
	MinTxFee   int    `json:"MinTxFee"`
	ActiveNet  string `json:"ActiveNet"`
}

type Configuration struct {
	Magic               uint32           `json:"Magic"`
	FoundationAddress   string           `json:"FoundationAddress"`
	Version             int              `json:"Version"`
	SeedList            []string         `json:"SeedList"`
	HttpRestPort        int              `json:"HttpRestPort"`
	MinCrossChainTxFee  int              `json:"MinCrossChainTxFee"`
	RestCertPath        string           `json:"RestCertPath"`
	RestKeyPath         string           `json:"RestKeyPath"`
	HttpInfoPort        uint16           `json:"HttpInfoPort"`
	HttpInfoStart       bool             `json:"HttpInfoStart"`
	OpenService         bool             `json:"OpenService"`
	HttpWsPort          int              `json:"HttpWsPort"`
	WsHeartbeatInterval time.Duration    `json:"WsHeartbeatInterval"`
	HttpJsonPort        int              `json:"HttpJsonPort"`
	OauthServerUrl      string           `json:"OauthServerUrl"`
	NoticeServerUrl     string           `json:"NoticeServerUrl"`
	NodePort            uint16           `json:"NodePort"`
	NodeOpenPort        uint16           `json:"NodeOpenPort"`
	PrintLevel          uint8            `json:"PrintLevel"`
	IsTLS               bool             `json:"IsTLS"`
	CertPath            string           `json:"CertPath"`
	KeyPath             string           `json:"KeyPath"`
	CAPath              string           `json:"CAPath"`
	MultiCoreNum        uint             `json:"MultiCoreNum"`
	MaxLogsSize         int64            `json:"MaxLogsSize"`
	MaxPerLogSize       int64            `json:"MaxPerLogSize"`
	MaxTxsInBlock       int              `json:"MaxTransactionInBlock"`
	MaxBlockSize        int              `json:"MaxBlockSize"`
	PowConfiguration    PowConfiguration `json:"PowConfiguration"`
}

type ConfigFile struct {
	ConfigFile Configuration `json:"Configuration"`
}

type ChainParams struct {
	Name               string
	PowLimit           *big.Int
	PowLimitBits       uint32
	TargetTimePerBlock time.Duration
	TargetTimespan     time.Duration
	AdjustmentFactor   int64
	MaxOrphanBlocks    int
	MinMemoryNodes     uint32
	CoinbaseLockTime   uint32
	ChainStorePath     string

	FoundationAddress1  string //20%
	FoundationAddress2  string //35%
	FoundationAddress3  string //17%
	FoundationAddress4  string //4.5%
	FoundationAddress5  string //4.5%
	FoundationAddress6  string //1%
	FoundationAddress7  string //2%
	FoundationAddress8  string //3%
	FoundationAddress9  string //5%
	FoundationAddress10 string //5%
	FoundationAddress11 string //3% reward
}

type configParams struct {
	*Configuration
	ChainParam *ChainParams
}

func init() {
	file, e := ioutil.ReadFile(DefaultConfigFilename)
	if e != nil {
		log.Fatalf("File error: %v\n", e)
		os.Exit(1)
	}
	// Remove the UTF-8 Byte Order Mark
	file = bytes.TrimPrefix(file, []byte("\xef\xbb\xbf"))

	config := ConfigFile{}
	e = json.Unmarshal(file, &config)
	if e != nil {
		log.Fatalf("Unmarshal json file erro %v", e)
		os.Exit(1)
	}
	//	Parameters = &(config.ConfigFile)
	Parameters.Configuration = &config.ConfigFile
	if Parameters.PowConfiguration.ActiveNet == "MainNet" {
		Parameters.ChainParam = mainNet
	} else if Parameters.PowConfiguration.ActiveNet == "TestNet" {
		Parameters.ChainParam = testNet
	} else if Parameters.PowConfiguration.ActiveNet == "RegNet" {
		Parameters.ChainParam = regNet
	}
}
